src/CollapsedForwarding.cc

```diff
@@ -146,6 +151,16 @@ CollapsedForwarding::HandleNotification(const Ipc::TypedMsgHdr &msg)
     HandleNewData("after notification");
 }
 
+/// Handle messages in IPC queue without relying on the queue state being in
```

Comment:
Please explicitly mention here which queue/messages will be "handled" in this manner.

Perhapse some parts of https://github.com/squid-cache/squid/pull/706#pullrequestreview-794271638 to make it clear why other worker process messages are (un)affected.

Reply:
Sorry I was quite vague on the request there.  The comment documents the *what*, however reading that info raises huge red flags. Such as why state matters at all if the queue can be used like this (surely its therefore safer to use the queue without the problematic state). Leaving one with worrisome implications -  "relying on" implies some major importance that in turn implies major risks of this function ignoring things.

My request is that you add to that comment, part of the clarifications you gave that explain the *why* such an apparently dangerous thing is both necessary and actually safe to do. I am not sure that I completely grasp it yet, but something like the following might ally the implied danger:
```
/// This method helps restores sanity of shared memory after worker death.
///
/// Handle messages in IPC queue without relying on the queue state
/// (such as number of unhandled messages)
/// being in sync with our state. Because the queue may be reflecting the state
/// of our killed predecessor.
```

Reply:
How about the new current version:

```C++
/// Handle queued IPC messages for the first time in this process lifetime, when
/// the queue may be reflecting the state of our killed predecessor.
```

The biggest problem here is that we want to avoid documenting how the function accomplishes its intended purpose (and even what specific queue states may be possible) because that information is irrelevant to the caller (i.e. to the primary function documentation audience) and because it is complicated/nuanced (i.e. cannot be accurately described in a couple of words). Today, the function resets the state of notification-related flags. Tomorrow, it may also set new/special flags to coordinate restart recovery with the message sender. All that is not what the caller needs/wants to know. Fucntion description should not change when those internal details change.

The caller only needs to know two or three things:

1. The function handles queued messages (the name kind of says that already, but we can detail it a bit to avoid arguing about the need for this function description in the first place).
2. This function must be called upon process startup (the name says that already, but we can detail it a bit to avoid arguing about the need for this function description in the first place).
3. Why the regular HandleNewData() call is not sufficient in this context (technically, this can be viewed as optional/obvious info but it may help developers understand why this special function was added with less thinking).

I will update the other function docs once these ones are settled.

----

> reading that info raises huge red flags. Such as why state matters at all if the queue can be used like this (surely its therefore safer to use the queue without the problematic state).

I do not think those thoughts are huge red flags. They are just questions about complex IPC queue functionality that go beyond any single queue-related method:

Q: Why not remove notifications completely if we can ignore their state at startup?
A: We would not know when to call HandleNewData() during runtime.

Q: Why not reset the whole notification state always/unconditionally, inside HandleNewData()?
A: Squid would send a lot more (slow/expensive) IPC messages. Worker-specific notification state is a notification-reducing optimization.

Q: Does that notification-reducing optimization add risks?
A: While any state, especially shared state, increases risks of bugs (like the one this PR is fixing), notification-reducing optimization does not reduce queue reliability as far as bug-free implementation is concerned.

Reply:
Okay. This reads much cleaner. Thank you for addressing this (and the explanations).